webhooks-with-pgstream.mdx

```diff
@@ -0,0 +1,555 @@
+
+![Postgres webhooks with pgstream](/images/pgstream-webhooks.png)
+
+## Preparation
+
+For this tutorial we need a Postgres instance that has logical replication enabled and the [wal2json](https://github.com/eulerto/wal2json) output plugin loaded. There are a number of ways to do this, but we're going to use the [docker-compose file](https://github.com/xataio/pgstream/blob/main/build/docker/docker-compose.yml) from the pgstream repo:
+
+```sh
+git clone https://github.com/xataio/pgstream.git
+```
+
+And then start the Postgres instance:
+
+```sh
+docker compose -f build/docker/docker-compose.yml up db
+```
+
+Then, in a new terminal window, you can connect to it and create a simple table to play with:
+
+```sh
+psql "postgres://postgres:postgres@localhost:5432/postgres"
+```
+
+At the psql prompt:
+
+```sql
+CREATE TABLE users(id SERIAL PRIMARY KEY, name TEXT, role TEXT);
+```
+
+Let's now install `pgstream`. One option is to build it from source, since we already have the repo cloned:
+
+```sh
+go build
+```
+
+The above assumes you have Go installed. If you don't, it might be easier to use the Homebrew tap:
+
+```sh
+brew tap xataio/pgstream
+brew install pgstream
+```
+
+Or you can find pre-built binaries on the [release pages](https://github.com/xataio/pgstream/releases).
+
+Next step is to initialize `pgstream`, like this:
+
+```sh
+pgstream init --pgurl "postgres://postgres:postgres@localhost/postgres?sslmode=disable"
+```
+
+This creates several tables managed by pgstream, which are grouped in a schema called `pgstream`.
+
+## Configuration
+
+The `pgstream` repo comes with a `pg2webhook.env` configuration file for this purpose. Its contents are:
+
+```sh
+# Listener config
+PGSTREAM_POSTGRES_LISTENER_URL="postgres://postgres:postgres@localhost?sslmode=disable"
+
+# Processor config
+PGSTREAM_TRANSLATOR_STORE_POSTGRES_URL="postgres://postgres:postgres@localhost?sslmode=disable"
+PGSTREAM_WEBHOOK_SUBSCRIPTION_STORE_URL="postgres://postgres:postgres@localhost?sslmode=disable"
+PGSTREAM_WEBHOOK_SUBSCRIPTION_STORE_CACHE_ENABLED=false
+PGSTREAM_WEBHOOK_SUBSCRIPTION_STORE_CACHE_REFRESH_INTERVAL="60s"
+```
+
+In this file, there are two Postgres connection strings defined. The first one is for the listener, to which database should pgstream connect and listen for changes. The second one is used as a metadata db to store the registered subscriptions and their configuration. In this example, they are both set to the same database, but you can use different databases or instances if you prefer.
+
+You are now ready to start `pgstream` like this:
+
+```sh
+pgstream run -c pg2webhook.env
+```
+
+## The webhooks listener
+
+We now have pgstream ready to go, but we need something to receive the webhooks. Let's use this simple python program (courtesy of ChatGPT) to start a webserver that logs all the requests it receives:
```

Comment:
It might be worth mentioning there is a webhook dummy server in the pgstream repository. It can be useful to have something readily available, and it can serve as a template to iterate on a custom implementation.
https://github.com/xataio/pgstream/blob/main/tools/webhook/webhook_server.go

Reply:
The reason for which I went with the python server is that the Go one didn't print the JSON as is, but rather the decoded messages. And I wanted to show the JSONs.

Reply:
Yeah, I was wondering why the example is in Python, it built the pgstream with Go, I somehow assumed examples will follow in Go.

Reply:
I'm also happy to update the go dummy server to show the json if that helps.

Reply:
I think, it would make the whole thing more coherent 🫶

Reply:
I have opened a PR with the changes:
https://github.com/xataio/pgstream/pull/64

Reply:
Awesome, thanks! Incorporated in the blog post.